packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/InternalTypedRealm.kt

```diff
@@ -34,4 +38,67 @@ internal interface InternalTypedRealm : TypedRealm {
+        // Be able to inject a cache here as well, so the Iterable case can share the cache
+        if (!obj.isManaged()) {
+            throw IllegalArgumentException("This object is unmanaged. Only managed objects can be copied: $obj.")
+        }
+        if (!obj.isValid()) {
+            throw IllegalArgumentException(
+                "Only valid objects can be copied from Realm. " +
+                    "This object was either deleted or the Realm has been closed, making this " +
+                    "object invalid: $obj."
+            )
+        }
+        if (depth < 0) {
+            throw IllegalArgumentException("Only a depth of 0 or more is allowed. Depth was: $depth.")
+        }
+        if (obj is RealmObjectInternal) {
+            val objectRef: RealmObjectReference<out BaseRealmObject> = obj.io_realm_kotlin_objectReference!!
+            val realmRef: RealmReference = objectRef.owner
+            val mediator: Mediator = realmRef.owner.configuration.mediator
+            return createDetachedCopy(mediator, obj, 0, depth, closeAfterCopy, cache)
+        } else {
+            throw IllegalStateException()
+        }
+    }
+
+    override fun <T : TypedRealmObject> copyFromRealm(obj: T, depth: Int, closeAfterCopy: Boolean): T {
+        return copyObjectFromRealm(obj, depth, closeAfterCopy, mutableMapOf())
+    }
+    override fun <T : TypedRealmObject> copyFromRealm(collection: Iterable<T>, depth: Int, closeAfterCopy: Boolean): List<T> {
+        val (valid: Boolean, nativePointer: NativePointer<*>?) = when (collection) {
+            is ManagedRealmList -> Pair(collection.isValid(), collection.nativePointer)
+            is ManagedRealmSet -> Pair(collection.isValid(), collection.nativePointer)
+            is RealmResultsImpl -> Pair(!collection.realm.isClosed(), collection.nativePointer)
+            else -> Pair(true, null)
+        }
+        if (!valid) {
+            throw IllegalArgumentException(
+                "Only valid collections can be copied from Realm. " +
+                    "This collection was either deleted or the Realm has been closed, making this " +
+                    "collection invalid"
+            )
+        }
+        val cache: ManagedToUnmanagedObjectCache = mutableMapOf()
+        return if (collection is Collection) {
+            // For collections we can pre-allocate the output array
+            val iter: Iterator<T> = collection.iterator()
+            MutableList(collection.size) { i: Int ->
+                copyObjectFromRealm(iter.next(), depth, closeAfterCopy, cache)
+            }
+        } else {
+            // Else we need to just do the naive approach
+            val result = ArrayList<T>()
+            collection.forEach { obj: T ->
+                val copiedObj = copyObjectFromRealm(obj, depth, closeAfterCopy, cache)
+                result.add(copiedObj)
+            }
+            result
+        }.also {
+            // If
```

Comment:
Shouldn't we release if the flag is set only (maybe why you left the comment above)?